action.yml

```
@@ -25,38 +19,58 @@ inputs:
 outputs:
   python-version:
     description: "Python version"
-    value: ${{ inputs.python_version }}
+    value: ${{ steps.setup-python.outputs.python-version }}
+  poetry-version:
+    description: "Poetry version"
+    value: ${{ steps.detect-versions.outputs.poetry-version }}
 
 runs:
   using: "composite"
   steps:
-    - uses: actions/setup-python@v4
-      with:
-        python-version: ${{ inputs.python_version }}
 
-    - name: Install and configure Poetry
+    - name: Determine Python and Poetry versions
+      id: detect-versions
       shell: bash
+      working-directory: ${{ inputs.working_directory }}
       run: |
-        curl -sSL https://install.python-poetry.org | python - --yes --version=${{ inputs.poetry_version }}
-        echo "$HOME/.poetry/bin" >> $GITHUB_PATH
+        if [ -z "${{ inputs.python_version }}" ] ; then
+            PYTHON_VERSION="$(sed -n -e '/^\[metadata\]/,/^\[/p' poetry.lock | sed -n -e 's/^python-versions[[:space:]]*=[[:space:]]*//p' | tr -d \"'[:space:]'\'~^)"
+        else
+            PYTHON_VERSION="${{ inputs.python_version }}"
+        fi
 
-    - name: Restore virtualenv from Poetry cache
-      uses: actions/cache@v3
-      id: poetry-cache
+        if [ -z "${{ inputs.poetry_version }}" ] ; then
+            POETRY_VERSION="$(head --lines=1 poetry.lock | grep --only-matching '[0-9]\+\.[0-9]\+\.[0-9]\+')"
+        else
+            POETRY_VERIONS="${{ inputs.poetry_version }}"
+        fi
+
+        echo "python-version=$PYTHON_VERSION" >> $GITHUB_OUTPUT
+        echo "poetry-version=$POETRY_VERSION" >> $GITHUB_OUTPUT
+
+    - name: Install and configure Poetry
+      shell: bash
+      run: pipx install poetry==${{ steps.detect-versions.outputs.poetry-version }}
+
+    - uses: actions/setup-python@v4
+      id: setup-python
       with:
-        path: ${{ inputs.poetry_virtualenvs_path }}
-        key: poetry-${{ inputs.python_version }}-${{ hashFiles('**/poetry.lock', '**/pyproject.toml') }}
+        python-version: '${{ steps.detect-versions.outputs.python-version }}'
+        cache: 'poetry'
 
-    - run: poetry lock --check
+    - run: |
+        # ToDo: can be removed somewhen in the future. This is for backwards compatibility for our current update to Poetry 1.6.1
+        if [ "x${{ steps.detect-versions.outputs.poetry-version }}" == "x1.5.1" ]; then
```

```suggestion
        if [ "$(echo ${{ steps.detect-versions.outputs.poetry-version }} | cut -c1-4)" == "1.5." ]; then
```

I think the version can't be empty, because the installation will fail first.

---

Updated and removed the todo.